### PR TITLE
Generic resource definitions

### LINKED ISSUE #806: Generic resource definitions
### Current situation
The current way how resource definitions work is that JADNC provides a generic base class, which users can derive from for a specific resource type. JADNC contains scanning logic that picks up such non-generic (resource-bound) classes and calls into them from various places.

Example:

```c#
public class OrderLineDefinition : ResourceDefinition<OrderLine>
{
    // ...
}
```

When no resource-bound class is found for a specific resource, then resource callbacks are skipped for that resource type. This is accomplished via a non-generic `IResourceDefinition` interface, which is a type intended for internal use only to track which callbacks are available. It put me on the wrong track when I started using JADNC: I expected this was something for me to implement (similar to services and repositories), but it is not.

### Proposal

I'd like to make resource definitions a first-class pluggable extensibility point, similar to services and repositories, so that resource definitions are *always* called into. We would provide an interface along with a default generic implementation (that contains no logic). Users can then implement the interface themselves and register their generic type -and- derive from our default generic implementation and register that -and- declare resource-bound definitions, which may derive from their own generic base class. 

Example:
```c#
namespace JADNC
{
    public interface IResourceDefinition<TResource>
    {
        // ...
    }

    public class JsonApiResourceDefinition<TResource> : IResourceDefinition<TResource>
    {
        // ...
    }
}

namespace ApiDeveloper
{
    public class CustomResourceDefinition<TResource> : JsonApiResourceDefinition<TResource>
    {
        // ...
    }

    public class OrderLineResourceDefinition : CustomResourceDefinition<OrderLine>
    {
        // ...
    }
}
```

### Use case

The use case I have in mind for this is authorization. Assume the system stores whether users have read/write access per resource type. While it is possible to annotate controller methods with attributes, this would not be sufficient. It would only work for primary requests like `/articles` and `/articles/1`, but not secondary requests like `/articles/1/author`. The controller would need to contain custom logic that inspects the `relationship` parameter and runs the permission check for that too. It is even debatable whether a permission check on articles should be done in that case, because the request is not targeting an article.

But it gets worse: a request like `/articles?include=author` would not get checked for access to the author resource. The current way to deal with that would be to override `ResourceDefinition.OnApplyIncludes` for all resource types in the project. A similar override for `ResourceDefinition.OnApplyFilter` is needed to properly check a request like `/articles?filter=any(revisions)` to ensure the caller has read permission on the revisions table.

If it were possible to plug in a generic resource definition, it could handle the permission check for all resource types in the project. Resource-bound classes would then call into the base class to ensure authorization is always applied.

### PR DESCRIPTION
**Pull request by bart-degreed**
Fixes #806.

Made resource definitions a first-class pluggable extensibility point, similar to resource services and repositories. So developers now can derive from `JsonApiResourceDefinition`, implement `IResourceDefinition`, or insert base classes in the type hierarchy to share common logic. Such a base class can then be registered as an open generic, so that all resources will use that (unless the container finds a non-generic class, which is a better match) instead of the built-in `JsonApiResourceDefinition`.

To accomplish this, I had to split up `ResourceDefinition` into two:
- `JsonApiResourceDefinition` is highly pluggable like described above
- `ResourceHooksDefinition` contains just the hook callbacks

This way, the recursive resolve logic for resource hooks remains as-is. Since it did not use `IResourceDefinitionProvider`, that type has been replaced with `IResourceDefinitionAccessor`, which invokes the requested callbacks directly on the found class. What used to be `IResourceDefinition` (an internal type to invoke callbacks on) no longer exists.

Additionally replaced `IHasMeta` with a method on `IResourceDefinition` and only scan for hook definitions when they are enabled. I intend to update documentation on `IHasMeta` (which was already wrong) later, as part of #737 (which may change these again).

### CODE REVIEW

File: src/JsonApiDotNetCore/Configuration/ServiceDiscoveryFacade.cs [hunk 1]
```diff
@@ -96,12 +101,12 @@ internal void DiscoverResources()
...
                 {
-                    AddResource(assembly, descriptor);
+                    AddResource(descriptor);
                 }
             }
         }
-        
-        internal void DiscoverInjectables()
+
+        internal void DiscoverInjectables(bool enableResourceHooks)
```
Comment (maurei) [inline src/JsonApiDotNetCore/Configuration/ServiceDiscoveryFacade.cs:109]: Alternatively we could pass `IJsonApiOptions` to the constructor of `ServiceDiscoveryFacade` and read the option from there. That feels a little bit cleaner to me. Thought?

**Review by maurei — CHANGES_REQUESTED** [review]
Partial review, will continue later today

File: src/JsonApiDotNetCore/Configuration/ServiceDiscoveryFacade.cs (hunk 1 above)
Comment (bart-degreed) [reply src/JsonApiDotNetCore/Configuration/ServiceDiscoveryFacade.cs]: Sounds good, I've change it.

**Review by maurei — APPROVED** [review]

File: docs/usage/meta.md [hunk 2]
```diff
@@ -1,43 +1,60 @@
 # Metadata
 
-Non-standard metadata can be added to your API responses in two ways: Resource and Request Meta. In the event of a key collision, the Request Meta will take precendence.
+Top-level custom metadata can be added to your API responses in two ways: globally and per resource type. In the event of a key collision, the resource meta will take precendence.
```
Comment (maurei) [inline docs/usage/meta.md:3]: I think the word "custom" is redundant here

File: docs/usage/meta.md [hunk 3]
```diff
@@ -1,43 +1,60 @@
...
 
-## Resource Meta
+## Global Meta
 
-Resource Meta is metadata defined on the resource itself by implementing the `IHasMeta` interface.
+Global metadata can be added by injecting a service that implements `IResponseMeta`.
```
Comment (maurei) [inline docs/usage/meta.md:7]: Is injecting the right terminology here? I usually refer to _injecting_ as adding a service to a constructors signature. Did you mean _registering_?

File: docs/usage/meta.md [hunk 4]
```diff
@@ -1,43 +1,60 @@
...
+This is useful if you need access to other injected services to build the meta object.
```
Comment (maurei) [inline docs/usage/meta.md:8]: _registered_ services

**Review by maurei — CHANGES_REQUESTED** [review]

File: docs/usage/meta.md (hunk 2 above)
Comment (bart-degreed) [reply docs/usage/meta.md:3]: It is, thanks.

File: docs/usage/meta.md (hunk 3 above)
Comment (bart-degreed) [reply docs/usage/meta.md:7]: Agreed.

File: docs/usage/meta.md (hunk 4 above)
Comment (bart-degreed) [reply docs/usage/meta.md:8]: Agreed.

**Review by maurei — APPROVED** [review]

**Pull request merged**